timelock-impl/src/test/java/com/palantir/atlasdb/timelock/paxos/ManualBatchingPingableLeaderTest.java

```diff
@@ -0,0 +1,174 @@
+        Future<LeaderPingResult> clientWhoIsNotLed =
+                manualBatchingPingableLeader.ping(HOST_UUID, CLIENT_WHO_IS_NOT_LED);
+
+        when(batchPingableLeader.ping(ImmutableSet.of(CLIENT_WHO_IS_LED, CLIENT_WHO_IS_NOT_LED)))
+                .thenReturn(ImmutableSet.of(CLIENT_WHO_IS_LED));
+
+        manualBatchingPingableLeader.runOneIteration();
+
+        assertThat(Futures.getUnchecked(clientWhoIsLed))
+                .isEqualTo(LeaderPingResults.pingReturnedTrue(HOST_UUID, HOST_AND_PORT));
+
+        assertThat(Futures.getUnchecked(clientWhoIsNotLed))
+                .isEqualTo(LeaderPingResults.pingReturnedFalse());
+    }
+
+    @Test
+    public void getUnresolvedFutureUntilSingleIterationHasRun() {
+        ManualBatchingPingableLeader manualBatchingPingableLeader = new ManualBatchingPingableLeader(
+                pingerWithContext,
+                Duration.ofSeconds(1),
+                Duration.ofSeconds(1),
+                HOST_UUID);
+
+        Future<LeaderPingResult> clientWhoIsLed = manualBatchingPingableLeader.ping(HOST_UUID, CLIENT_WHO_IS_LED);
+
+        assertThat(clientWhoIsLed)
+                .as("it's hard to test that it never gets resolved prior to running the single iteration")
+                .isNotDone();
+
+        when(batchPingableLeader.ping(ImmutableSet.of(CLIENT_WHO_IS_LED)))
+                .thenReturn(ImmutableSet.of(CLIENT_WHO_IS_LED));
+
+        manualBatchingPingableLeader.runOneIteration();
+
+        assertThat(clientWhoIsLed)
+                .as("we expect this to be done, everything is running within a single thread in this test")
+                .isDone();
+    }
+
+    @Test
+    public void getBackCachedResultIfTryingToPingTwice() {
+        ManualBatchingPingableLeader manualBatchingPingableLeader = new ManualBatchingPingableLeader(
+                pingerWithContext,
+                Duration.ofSeconds(1),
+                Duration.ofSeconds(1),
+                HOST_UUID);
+
+        Future<LeaderPingResult> clientWhoIsLed = manualBatchingPingableLeader.ping(HOST_UUID, CLIENT_WHO_IS_LED);
+
+        when(batchPingableLeader.ping(ImmutableSet.of(CLIENT_WHO_IS_LED)))
+                .thenReturn(ImmutableSet.of(CLIENT_WHO_IS_LED))
+                .thenReturn(ImmutableSet.of());
+
+        manualBatchingPingableLeader.runOneIteration();
+
+        assertThat(Futures.getUnchecked(clientWhoIsLed))
+                .isEqualTo(LeaderPingResults.pingReturnedTrue(HOST_UUID, HOST_AND_PORT));
+
+        Future<LeaderPingResult> clientWhoIsLedSecondRequest =
+                manualBatchingPingableLeader.ping(HOST_UUID, CLIENT_WHO_IS_LED);
+
+        assertThat(Futures.getUnchecked(clientWhoIsLedSecondRequest))
+                .isEqualTo(LeaderPingResults.pingReturnedTrue(HOST_UUID, HOST_AND_PORT));
+
+        verify(batchPingableLeader, only()).ping(ImmutableSet.of(CLIENT_WHO_IS_LED));
+    }
```

Comment:
There's a test that I think is quite important that's missing here:
- `runOneIteration` will ping (and continue to ping) services that haven't been requested since the last iteration - this is the behaviour I was confused by, but given its usage I think would be good to codify it